design/rfc-01-fwu-suit/rfc-01.rst

```diff
@@ -0,0 +1,693 @@
+
+Complex installation sequences
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Where a SUIT manifest includes additional payload fetching as part of the suit-install command sequence, the manifest processing and payload fetching cycle would be repeated after the installation commences. This might occur immediately if no reboot is required, or as part of a dedicated Bootloader/Installer state of the system if this requires a system restart.
+
+`Figure 2 <fig-install-flow_>`_ shows an example of the additional flow when fetching during installation after a reboot.
+
+.. _fig-install-flow:
+
+.. figure:: suit-install.svg
+
+   **Figure 2** *The additional flow in a complex SUIT installation*
+
+
+Draft API design
+----------------
+
+Envelope and payload transfer
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+`Figure 3 <fig-fetch_>`_ shows the detailed call sequence for the Update client when initially processing a SUIT envelope
+
+.. _fig-fetch:
+
+.. figure:: fetch-sequence.svg
+
+   **Figure 3** *The normal call sequence for initial SUIT update*
+
+   This shows the transfer of the envelope, and the processing of the suit-payload-fetch command sequence.
+
+The detailed steps of the flow are as follows (the numbers refer to the corresponding number in Figure 3):
+
+*  1: The process is started by transferring the SUIT envelope as a firmware image using a component identifier allocated to the SUIT envelope component.
+
+*  6: The call to ``psa_fwu_finish()`` behaves differently when processing a SUIT envelope. On a successful transfer, the call will return a new response code, ``PSA_FWU_PROCESSING_REQUIRED``, to indicate that the component requires processing. At this point the envelope component will be in a new ``PSA_FWU_FETCHING`` state, instead of the typical ``PSA_FWU_CANDIDATE`` state.
+
+*  7: If processing is required, the Update client then calls ``psa_fwu_process()`` to begin manifest processing. At this point the Update service will do the following:
+
+   * Verify and authenticate the manifest.
+   * Process the Update command sequences: system validation, dependency resolution, payload fetch, payload verification.
+
+*  8: If a payload is required that is detached from the Envelope, the call to ``psa_fwu_process()`` returns with a new status code, ``PSA_FWU_PAYLOAD_REQUIRED``. The call to ``psa_fwu_process()`` includes an output parameter, which the service uses to provide the details of the payload to be transferred, including a component identifier for the payload (for use with ``psa_fwu_start()`` etc), and a URI for the payload.
+
+   *The Update service might also have information about the size and digest of the payload to be fetched. These could be optionally be provided with the payload URI: are there benefits in using this information to eliminate or detect incorrect or malicious transfers prior to transfer to the Update service?*
```

Comment:
That functionality may cause confusions. In SUIT, a digest value (or expected payload size) in not a fixed property of component but rather an argument to directive-override-parameters ‘, so:
-	Expected digest value (or expected payload size) may be unknown at time of execution of ‘directive-fetch‘ – there is nothing what enforces a manifest creator to call ‘directive-override-parameters’ prior to ‘directive-fetch‘
-	Manifest may be crafted in a way that it accepts alternative payloads represented by single URI. In such case manifest may evaluate several digest values on fetched payload.
-	Methods of digest calculation for regular image and dependency manifest differs.
-	Manifest may be crafted in a way that it repeats fetching in case if recently downloaded payload is damaged, so some resiliency mechanisms are already in-place.

Reply:
It is not unlikely that the digest parameter will already be set when calling suit-directive-fetch:

- One of the optimisations suggested for a SUIT processor when restarting a command sequence after interruption, is to elide the fetching of a payload that has already been transferred, by matching a stored payload using the digest of the requested payload.

- This optimisation depends on the component's digest parameter being set prior to the suit-directive-fetch command. In the absence of the digest parameter the optimisation is disabled, and the payload will always be fetched.

I can see that for some systems, being able to detect a non-matching image before download, might avoid costly download or store-to-flash overheads. This could improve long-term-reliability, but only if having incorrect payload images on the update server is considered to be probable. It does not significantly mitigate security concerns, as an attacker can present an acceptable digest, but an invalid image, and would still result in the download and transfer of the image before the SUIT digest verification reports an error.

### Conclusion

Adding these _might_ be beneficial for some systems/deployments, but the availability of these attributes is not guaranteed. Therefore, _if_ we add them to the payload information, then they must be optional, and we need another field in the payload information to individually indicate if these parameters are present in the information.

Reply:
This is now (from v0.8 of the proposal) optionally provided in the payload information, with a flag to indicate its presence.